### ISSUE TITLE
Add an immutable SortedSet to the javaslang.collection package

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#214]

**Comment by danieldietrich** [comment]
A `SortedSet<T extends Comparable<T>>` consists of objects which are totally ordered. Here, we will back it by a `RedBlackTree` which is currently one of the most efficient binary search tree data structures.

A general `Set<T>` consists of arbitrary objects which can be compared using equals(). Currently I'm investigating how to implement a `Set` in the most efficient way. E.g. the set could be backed by a RedBlackTree containing elements `(hashCode, List)` to simulate a `HashMap`. Maybe a `Trie` or a `Map` is more efficient, depending on how many hash collisions are expected to occur etc. This has to be further investigated...

**Comment by danieldietrich** [comment]
Further reading on how to delete nodes of a RedBlackTree: https://fstaals.net/RedBlackTree.html

**Comment by danieldietrich** [comment]
For delete we need the same rotate methods as for insert:

```java
    // We target in all cases:
    //
    //     (Ry)
    //     /   \
    //  (Bx)   (Bz)
    //  /  \   /  \
    // a    b c    d
    //
    // = (R:y (B:x R:a R:b) (B:z R:c R:d))
    static final TreeNode<Character> EXAMPLE = red(black(red('a'), 'x', red('b')), 'y', black(red('c'), 'z', red('d')));

    /**
     * Case 0:
     * <pre><code>
     *        (Bz)
     *        /  \
     *     (Ry)   d
     *     /  \
     *  (Rx)   c
     *  /  \
     * a    b
     * </code></pre>
     */

    /**
     * Case 1:
     * <pre><code>
     *     (Bz)
     *     /  \
     *   (Rx)  d
     *  /   \
     * a    (Ry)
     *      /  \
     *     b    c
     * </code></pre>
     */

    /**
     * Case 2:
     * <pre><code>
     *  (Bx)
     *  /  \
     * a   (Rz)
     *     /  \
     *   (Ry)  d
     *   /  \
     *  b    c
     * </code></pre>
     */

    /**
     * <strong>Case 3:</strong>
     *
     * <pre><code>
     *  (Bx)
     *  /  \
     * a   (Ry)
     *     /  \
     *    b   (Rz)
     *        /  \
     *       c    d
     * </code></pre>
     */
```

**Comment by danieldietrich** [comment]
[Matt Might's delete algorithm]() is considered to be [simpler to understand](http://matt.might.net/articles/red-black-delete) than the one of Stefan Kahrs ("Red-black trees with types", Journal of functional programming, 11(04), pp 425-432, July 2001).

[A benchmark](http://matt.might.net/papers/germane2014deletion.pdf) showed, that Kahrs version is more efficient.

[Andrew Appel’s version](https://www.cs.princeton.edu/~appel/papers/redblack.pdf) of Kahrs should be taken into account, too.

**Comment by danieldietrich** [comment]
I found a [Haskell version of Stefan Kahr's algorithm](https://github.com/danieldietrich/llrbtree/blob/master/Data/Set/RBTree.hs), implemented 2012 by Kazu Yamamoto.

The RedBlackTree has the following operations:

```haskell
  -- * Creating red-black trees
  , empty
  , singleton
  , insert
  , fromList
  -- * Converting to a list
  , toList
  -- * Membership
  , member
  -- * Deleting
  , delete
  , deleteMin
  , deleteMax
  -- * Checking
  , null
  -- * Set operations
  , union
  , intersection
  , difference
  -- * Helper functions
  , join
  , merge
  , split
  , minimum
  , maximum
  , valid
  , showSet
  , printSet
```

**Comment by danieldietrich** [comment]
Fixed with #545, #585 and #605.
